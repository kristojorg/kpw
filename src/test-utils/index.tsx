import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ThemeProvider } from "theme-ui";
import theme from "../theme";
import ContextProvider from "../components/context/ContextProvider";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import library from "./fixtures/library";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import BasicAuthPlugin from "../auth/basicAuthPlugin";
import buildStore from "opds-web-client/lib/store";
import { State } from "opds-web-client/lib/state";
import { LibraryData } from "../interfaces";
import "./mockScrollTo";
import * as fixtures from "./fixtures";
import ActionsCreator from "opds-web-client/lib/actions";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import { adapter } from "opds-web-client/lib/OPDSDataAdapter";
import serializer from "jest-emotion";

expect.addSnapshotSerializer(serializer);

/**
 * uncomment this if you would like console errors to to
 * error the test. Useful if you don't know where a console
 * error is coming from.
 */
// const consoleErrorSpy = jest.spyOn(global.console, "error");
// consoleErrorSpy.mockImplementation((msg, ...opts) => {
//   console.warn(msg, ...opts);
//   throw new Error(msg);
// });

export { fixtures };

configure({ adapter: new Adapter() });

/**
 * mock out the window.URL.createObjectURL since it isn't
 * available on jsdom
 */
const mockCreateObjectURL = jest.fn();
global["URL"].createObjectURL = mockCreateObjectURL;

/**
 * We create the actions and fetcher here so that they can
 * be imported by our test file and spied on using:
 * jest.spyOn(actions, "fetchSearchDescription")
 * and then asserted against
 */
export const fetcher = new DataFetcher({ adapter });
export const actions = new ActionsCreator(fetcher);

type CustomRenderOptions = Parameters<typeof render>[1] & {
  route?: string;
  initialState?: State;
  library?: LibraryData;
};
const customRender = (ui: any, options?: CustomRenderOptions) => {
  const pathFor = jest
    .fn()
    .mockImplementation((collectionUrl?: string, bookUrl?: string) =>
      `/${collectionUrl}` + bookUrl ? `/${bookUrl}` : ""
    );

  const store = buildStore(options?.initialState, [BasicAuthPlugin], pathFor);

  // spy on dispatch by default
  const origDispatch = store.dispatch;
  const mockDispatch: jest.Mock<typeof origDispatch> = jest
    .fn()
    .mockImplementation(origDispatch);
  store.dispatch = mockDispatch as typeof origDispatch;
  // const dispatch = jest.spyOn(store, "dispatch");

  const history = createMemoryHistory({
    initialEntries: [options?.route ?? "/"]
  });
  const AllTheProviders = ({ children }) => {
    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <ContextProvider
            library={options?.library ?? library}
            shortenUrls
            helmetContext={{}}
            initialState={options?.initialState}
            store={store}
            fetcher={fetcher}
            actions={actions}
          >
            {children}
          </ContextProvider>
        </ThemeProvider>
      </Router>
    );
  };

  return {
    ...render(ui, { wrapper: AllTheProviders, ...options }),
    // we pass our mocks along so they can be used in assertions
    store,
    history,
    dispatch: mockDispatch
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
