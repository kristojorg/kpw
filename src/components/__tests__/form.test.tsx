import * as React from "react";
import { render } from "../../test-utils";
import FormInput from "../form/FormInput";

test("renders as expected", () => {
  const node = render(<FormInput name="input-name" label="Email" />);
  expect(node.container.firstChild).toMatchSnapshot();
});

test("displays star when required", () => {
  const node = render(<FormInput required name="input-name" label="Email" />);
  expect(node.getByText("*")).toBeInTheDocument();
});

test("no star when not required", () => {
  const node = render(<FormInput name="input-name" label="Email" />);
  expect(node.queryByText("*")).toBeNull();
});

test("displays label", () => {
  const node = render(<FormInput name="input-name" label="Email" />);
  expect(node.getByText("Email")).toBeInTheDocument();
});

test("shows error when present", () => {
  const node = render(
    <FormInput name="input-name" label="Email" error="Something went wrong" />
  );
  expect(node.getByText("Something went wrong")).toBeInTheDocument();
});
