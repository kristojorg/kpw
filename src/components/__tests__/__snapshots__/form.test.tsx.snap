// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
.emotion-2 {
  margin-bottom: 8px;
}

.emotion-0 {
  margin-bottom: 4px;
  display: inline-block;
}

.emotion-1 {
  border-radius: 4px;
  border: 1px solid;
  border-color: var(--theme-ui-colors-primary,#0467a6);
  border-width: 2px;
  padding: 4px;
  font-size: 16px;
  display: block;
}

<div
  class="emotion-2"
  role="group"
>
  <label
    class="emotion-0"
    for="input-name"
  >
    Email
  </label>
  <input
    aria-describedby="input-name-errors"
    aria-label="input-name input"
    class="emotion-1"
    id="input-name"
    name="input-name"
    type="text"
  />
</div>
`;
